mass_mailing_company_newsletter/models/res_partner.py

```diff
@@ -0,0 +1,191 @@
+        help="Technical field: The company's newsletter subscription for this partner.",
+        compute="_compute_main_mailing_list_subscription_id",
+        compute_sudo=True,
+        search="_search_main_mailing_list_subscription_id",
+    )
+    main_mailing_list_subscription_state = fields.Selection(
+        string="Company Newsletter Subscription State",
+        selection=[
+            ("subscribed", "Subscribed"),
+            ("unsubscribed", "Unsubscribed"),
+        ],
+        compute="_compute_main_mailing_list_subscription_state",
+        compute_sudo=True,
+        inverse="_inverse_main_mailing_list_subscription_state",
+        search="_search_main_mailing_list_subscription_state",
+        tracking=True,
+    )
+    main_mailing_list_subscription_date = fields.Datetime(
+        related="main_mailing_list_subscription_id.subscription_date",
+        string="Company Newsletter Subscription Date",
+    )
+    main_mailing_list_unsubscription_date = fields.Datetime(
+        related="main_mailing_list_subscription_id.unsubscription_date",
+        string="Company Newsletter Unsubscription Date",
+    )
+
+    @api.depends("company_id")
+    @api.depends_context("company")
+    def _compute_main_mailing_list_id(self):
+        self.main_mailing_list_id = self.env.company.main_mailing_list_id
+
+    @api.depends("mailing_contact_id")
+    @api.depends_context("company")
+    def _compute_main_mailing_list_subscription_id(self):
+        if not self.main_mailing_list_id or not self.mailing_contact_id:
+            self.main_mailing_list_subscription_id = False
+            return
+        # Find the mailing.contact.subscription
+        subs = self.env["mailing.contact.subscription"].search(
+            [
+                ("contact_id", "in", self.mailing_contact_id.ids),
+                ("list_id", "=", self.main_mailing_list_id.id),
+            ]
+        )
+        subs_by_contact_id = {sub.contact_id.id: sub.id for sub in subs}
+        for rec in self:
+            rec.main_mailing_list_subscription_id = subs_by_contact_id.get(
+                rec.mailing_contact_id.id
+            )
+
+    def _search_main_mailing_list_subscription_id(self, operator, value):
+        # Functionally speaking, this shouldn't be necessary.
+        # But without it, Odoo would print an error log saying:
+        # Non-stored field main_mailing_list_subscription_id cannot be searched.
+        # Even though we're not explicitly searching on it, the following line is:
+        # https://github.com/odoo/odoo/blob/c9b6b030f/odoo/models.py#L5860
+        if operator != "in":  # pragma: no cover
+            return NotImplementedError()
+        # Rely on the fact that search takes a query as value since Odoo 14.0
+        # See https://github.com/odoo/odoo/commit/69869ab68
+        mailing_list = self.env.company.main_mailing_list_id
+        search = self.env["mailing.contact.subscription"]._search(
```

Comment:
```suggestion
        ids = self.env["mailing.contact.subscription"]._search(
```

Reply:
As of https://github.com/odoo/odoo/commit/69869ab68, `_search` returns a `Query` object, not ids.

Reply:
then `query`? :smile: 
As you prefer.. not important ;)